nexus/src/app/background/tasks/tuf_artifact_replication.rs

```diff
@@ -0,0 +1,447 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! TUF Repo Depot: Artifact replication across sleds (RFD 424)
+//!
+//! `Nexus::updates_put_repository` accepts a TUF repository, which Nexus
+//! unpacks, verifies, and reasons about the artifacts in. This uses temporary
+//! storage within the Nexus zone, so the update artifacts have to go somewhere.
+//! We've settled for now on "everywhere": a copy of each artifact is stored on
+//! each sled's M.2 devices.
+//!
+//! This background task is responsible for getting locally-stored artifacts
+//! onto sleds, and ensuring all sleds have copies of all artifacts.
+//! `Nexus::updates_put_repository` sends the [`ArtifactsWithPlan`] object to
+//! this task via an [`mpsc`] channel and activates it.
+//!
+//! During each activation:
+//!
+//! 1. The task moves `ArtifactsWithPlan` objects of the `mpsc` channel and into
+//!    a `Vec`.
+//! 2. The task queries the list of artifacts stored on each sled, and compares
+//!    it to the list of artifacts in CockroachDB. Sled artifact storage is
+//!    content-addressed by SHA-256 checksum. Errors are logged but otherwise
+//!    ignored (unless no sleds respond); the task proceeds as if that sled
+//!    has no artifacts. (This means that the task will always be trying to
+//!    replicate artifacts to that sled until it comes back or is pulled out
+//!    of service.)
+//! 3. The task compares the list of artifacts from the database to the list of
+//!    artifacts available locally.
+//! 4. If all the artifacts belonging to an `ArtifactsWithPlan` object have
+//!    been replicated to at least `MIN_SLED_REPLICATION` sleds, the task drops
+//!    the object from its `Vec` (thus cleaning up the local storage of those
+//!    files).
+//! 5. The task generates a list of requests that need to be sent:
+//!    - PUT each locally-stored artifact not present on any sleds to a random
+//!      sled.
+//!    - For each partially-replicated artifact, choose a sled that is missing
+//!      the artifact, and tell it (via `artifact_copy_from_depot`) to fetch the
+//!      artifact from a random sled that has it.
+//!    - DELETE all artifacts no longer tracked in CockroachDB from all sleds
+//!      that have that artifact.
+//! 6. The task randomly choose requests up to per-activation limits and
+//!    sends them. Up to `MAX_REQUESTS` total requests are sent, with up to
+//!    `MAX_PUT_REQUESTS` PUT requests. Successful and unsuccessful responses
+//!    are logged.
```

Comment:
Current repositories have 41 artifacts (this is assuming the future work of splitting up the composite control plane artifact, because I want sled agent to be able to use these artifacts directly from the update dataset to create control plane zones). With `MAX_PUT_REQUESTS = 8`, it will take `ceil(41 / 8) = 6` executions (or 5 minutes from upload time) for one copy of every artifact to have made it to at least one sled.

Once all artifacts are present at least somewhere, all 3 Nexus instances will begin to run replication. This will take a total of `ceil(41 * 32 / 31) = 43` executions split across the 3 Nexus instances.

So the actual concrete amount of time it takes to fully replicate is about 20 minutes.

There's two reasons I'm throttling it this way.
- For PUT requests: it's definitely a worry about spiking the network bandwidth from the Nexus instance the repo happens to be present on to the rest of the rack.
- For other requests: the goal here was to prevent a sort of split-brain situation as described above, where different Nexus instances are running this task with different views of which artifacts ought to be present, by keeping overall task execution time brief.

I think if we implement a generation number for the state of the TUF repos table, and we aren't worried about bandwidth spikes here, we don't need either of the per-execution caps.